lib/rust/ensogl/component/grid-view/src/lib.rs

```diff
@@ -0,0 +1,508 @@
+#![warn(missing_debug_implementations)]
+#![warn(missing_docs)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unused_import_braces)]
+#![warn(unused_qualifications)]
+
+
+// ==============
+// === Export ===
+// ==============
+
+pub mod basic;
+pub mod entry;
+pub mod scrollable;
+pub mod visible_area;
+
+pub use ensogl_scroll_area::Viewport;
+
+
+
+/// Commonly used types and functions.
+pub mod prelude {
+    pub use ensogl_core::prelude::*;
+}
+
+use crate::prelude::*;
+
+use enso_frp as frp;
+use ensogl_core::application::command::FrpNetworkProvider;
+use ensogl_core::application::Application;
+use ensogl_core::display;
+use ensogl_core::display::scene::layer::WeakLayer;
+use ensogl_core::display::scene::Layer;
+use ensogl_core::gui::Widget;
+
+use crate::entry::EntryFrp;
+use crate::visible_area::all_visible_locations;
+use crate::visible_area::visible_columns;
+use crate::visible_area::visible_rows;
+pub use entry::Entry;
+
+
+// ===========
+// === FRP ===
+// ===========
+
+/// A row index in [`GridView`].
+pub type Row = usize;
+/// A column index  in [`GridView`].
+pub type Col = usize;
+
+ensogl_core::define_endpoints_2! {
+    <EntryModel: (frp::node::Data), EntryParams: (frp::node::Data)>
+    Input {
+        /// Declare what area of the GridView is visible. The area position is relative to left-top
+        /// corner of the Grid View.
+        set_viewport(Viewport),
+        /// Reset entries, providing number of rows and columns. All currently displayed entries
+        /// will be detached and their models re-requested.
+        reset_entries(Row, Col),
+        /// Probide model for specific entry. Should be called only after `model_for_entry_needed`
+        /// event for given row and column. After that the entry will be visible.
+        model_for_entry(Row, Col, EntryModel),
+        /// Set the entries size. All entries have the same size.
+        set_entries_size(Vector2),
+        /// Set the entries parameters.
+        set_entries_params(EntryParams),
+        /// Set the layer for any texts rendered by entries. The layer will be passed to entries'
+        /// constructors. **Performance note**: This will re-instantiate all entries.
+        set_text_layer(Option<WeakLayer>),
+    }
+
+    Output {
+        row_count(Row),
+        column_count(Col),
+        viewport(Viewport),
+        entries_size(Vector2),
+        entries_params(EntryParams),
+        content_size(Vector2),
+        /// Event emitted when the Grid View needs model for an uncovered entry.
+        model_for_entry_needed(Row, Col),
+    }
+}
+
+
+
+// =============
+// === Model ===
+// =============
+
+// === EntryCreationCtx ===
+
+#[derive(CloneRef, Debug, Derivative)]
+#[derivative(Clone(bound = ""))]
+struct EntryCreationCtx<P> {
+    app:              Application,
+    network:          frp::WeakNetwork,
+    set_entry_size:   frp::Stream<Vector2>,
+    set_entry_params: frp::Stream<P>,
+}
```

Comment:
Any of these is needed:
- expanding "P" to something named more obviously
- docs explaining what "P" os

Either way, the docs are needed. It's not obvious to me what this struct is by reading its name.